### PR TITLE
feat: Add `OPERATOR` tier

### LINKED ISSUE #3099: [HBAR Rate Limit Redesign] Add OPERATOR tier
Instead of tracking the remaining total budget in-memory inside `HbarLimitService` (we would have to divide it by the number of Kubernetes instances, etc - it is creating unnecessary complexity in the configurations), we could create a new subscription tier (`SubscriptionTier.OPERATOR`) and add a pre-configured spending plan in the JSON configuration file which links the ETH address(es) of the relay operator account(s) to it.

In this way we can only set the total limit for the `OPERATOR` tier to `HBAR_RATE_LIMIT_TINYBAR` and it will be shared by all instances of the relay.

### PR DESCRIPTION
**Pull request by victor-yanev**
**Description**:

Instead of tracking the remaining total budget in-memory inside `HbarLimitService` (we would have to divide it by the number of Kubernetes instances, etc - it is creating unnecessary complexity in the configurations), we could create a new subscription tier (`SubscriptionTier.OPERATOR`) and add a pre-configured spending plan in the JSON configuration file which links the EVM address(es) of the relay operator account(s) to it.

In this way we can only set the total limit for the `OPERATOR` tier to `HBAR_RATE_LIMIT_TINYBAR` and it will be shared by all instances of the relay.

**Related issue(s)**:

Fixes #3099

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by konstantinabl — APPROVED** [review]
lgtm! but please fix the failing tests in CI

File: packages/relay/src/lib/relay.ts [hunk 1]
```diff
@@ -160,13 +160,15 @@ export class RelayImpl implements Relay {
...
       register,
-      Hbar.fromTinybars(total),
       duration,
     );
 
     const hapiService = new HAPIService(logger, register, this.cacheService, this.eventEmitter, hbarLimitService);
 
     this.clientMain = hapiService.getMainClientInstance();
+    if (this.clientMain.operatorAccountId) {
+      hbarLimitService.setOperatorAddress(this.clientMain.operatorAccountId.toSolidityAddress());
```
Comment (Nana-EC) [inline packages/relay/src/lib/relay.ts:170]: Q: why is this needed when it's set in the `hbarLimitService` constructor?

File: packages/config-service/src/services/globalConfig.ts [hunk 2]
```diff
@@ -283,43 +283,31 @@ export class GlobalConfig {
       envName: 'HBAR_RATE_LIMIT_BASIC',
       type: 'number',
       required: false,
-      defaultValue: null,
+      defaultValue: 1_120_000_000, // 11.2 hbar
```
Comment (Nana-EC) [inline packages/config-service/src/services/globalConfig.ts:286]: nit: note the assumed duration for context

File: docs/configuration.md [hunk 3]
```diff
@@ -111,7 +111,6 @@ Unless you need to set a non-default value, it is recommended to only populate o
...
 | `SEND_RAW_TRANSACTION_SIZE_LIMIT`           | "131072"                              | Sets the limit of the transaction size the relay accepts on eth_sendRawTransaction                                                                                                                                                                                                                     |
 | `GET_RECORD_DEFAULT_TO_CONSENSUS_NODE`      | "false"                               | Flag to set if get transaction record logic should first query the mirror node (false) or consensus node via the SDK (true).                                                                                                                                                                           |
-| `HBAR_RATE_LIMIT_WHITELIST`                 | [""]                                  | An array of EVM addresses that are allowed to bypass the HBAR rate limits                                                                                                                                                                                                                              |
```
Comment (Nana-EC) [inline docs/configuration.md:114]: Q: Was this ever used?
If so what version was this added in?

File: packages/relay/src/lib/db/types/hbarLimiter/subscriptionTier.ts [hunk 4]
```diff
@@ -33,4 +33,9 @@ export enum SubscriptionTier {
    * Trusted Partners
    */
   PRIVILEGED = 'PRIVILEGED',
+
+  /**
+   * Relay Operators
+   */
+  OPERATOR = 'OPERATOR',
```
Comment (Nana-EC) [inline packages/relay/src/lib/db/types/hbarLimiter/subscriptionTier.ts:40]: I think this requires an update to the design doc.
Did this already happen or will it be done in follow up?

File: docs/configuration.md (hunk 3 above)
Comment (victor-yanev) [reply docs/configuration.md:114]: This is from the old hbar limiter and it's no longer used anywhere, so I removed it

File: packages/relay/src/lib/services/hbarLimitService/index.ts [hunk 5]
```diff
@@ -74,30 +77,37 @@ export class HbarLimitService implements IHbarLimitService {
...
     private readonly limitDuration: number,
   ) {
     this.reset = this.getResetTimestamp();
-    this.remainingBudget = this.totalBudget;
 
-    if (this.totalBudget.toTinybars().lte(0)) {
+    const operatorId = ConfigService.get('OPERATOR_ID_MAIN');
+    const operatorKey = ConfigService.get('OPERATOR_KEY_MAIN');
+    if (operatorId) {
+      this.operatorAddress = AccountId.fromString(operatorId as string).toSolidityAddress();
```
[the remark this answers is not in the document]
Comment (victor-yanev) [reply packages/relay/src/lib/services/hbarLimitService/index.ts:104]: Sure, we can do that

File: packages/server/tests/acceptance/hbarLimiter.spec.ts [hunk 6]
```diff
@@ -251,8 +251,19 @@ describe('@hbarlimiter HBAR Limiter Acceptance Tests', function () {
...
         await hbarSpendingPlanRepository.resetAmountSpentOfAllPlans(requestDetails);
 
+        // we don't want to reset the operator account's spending, so we have to add the amount spent back
+        for (const plan of operatorPlans) {
+          await hbarSpendingPlanRepository.addToAmountSpent(plan.id, plan.amountSpent, requestDetails, mockTTL);
+        }
+
         // Note: Since the total HBAR budget is shared across the entire Relay instance by multiple test cases,
         //       and expense updates occur asynchronously, the wait below ensures that the HBAR amount has sufficient time
         //       to update properly after each test.
```
[the remark this answers is not in the document]
Comment (victor-yanev) [reply packages/server/tests/acceptance/hbarLimiter.spec.ts:269]: Not really, the total limit tests already cover this

[the remark this answers is not in the document]
Comment (victor-yanev) [reply packages/server/tests/acceptance/hbarLimiter.spec.ts:269]: > https://github.com/hashgraph/hedera-json-rpc-relay/pull/3316#discussion_r1881668357

The operator plan essentially replaces the in-memory tracking of the total limit that we used to do before

File: packages/relay/src/lib/services/hbarLimitService/index.ts [hunk 7]
```diff
@@ -532,4 +546,42 @@ export class HbarLimitService implements IHbarLimitService {
...
+   * @private
+   */
+  private async getOperatorSpendingPlan(requestDetails: RequestDetails): Promise<IDetailedHbarSpendingPlan> {
+    let operatorPlan = await this.getSpendingPlan(this.operatorAddress!, requestDetails);
+    if (!operatorPlan) {
+      this.logger.trace(`${requestDetails.formattedRequestId} Creating operator spending plan...`);
+      operatorPlan = await this.createSpendingPlanForAddress(
+        this.operatorAddress!,
+        requestDetails,
+        SubscriptionTier.OPERATOR,
```
[the remark this answers is not in the document]
Comment (victor-yanev) [reply packages/relay/src/lib/services/hbarLimitService/index.ts:563]: We would still need this logic, so that the operator plan is recreated in case the cache expires (after the limit duration).

Otherwise, if we don't have this on-demand creation, we will need custom logic to create it in intervals of time which could create race conditions and the operator plan might not yet be created for incoming requests.

File: packages/relay/src/lib/services/hbarLimitService/index.ts [hunk 8]
```diff
@@ -74,30 +77,37 @@ export class HbarLimitService implements IHbarLimitService {
...
-  private remainingBudget: Hbar;
+  private reset: Date;
 
   /**
-   * The reset timestamp for the rate limiter.
+   * The operator address for the rate limiter.
    * @private
    */
-  private reset: Date;
+  private operatorAddress?: string;
```
[the remark this answers is not in the document]
Comment (victor-yanev) [reply packages/relay/src/lib/services/hbarLimitService/index.ts:89]: Yes, we can but we would need to duplicate the really screwed up logic in hapiService.ts:

https://github.com/hashgraph/hedera-json-rpc-relay/blob/main/packages/relay/src/lib/services/hapiService/hapiService.ts#L334-L359

File: packages/config-service/src/services/globalConfig.ts (hunk 2 above)
Comment (victor-yanev) [reply packages/config-service/src/services/globalConfig.ts:286]: It's right below it:
image

File: packages/relay/src/lib/services/hbarLimitService/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (victor-yanev) [reply packages/relay/src/lib/services/hbarLimitService/index.ts:89]: I opened another [draft PR](https://github.com/hashgraph/hedera-json-rpc-relay/pull/3331) which fixes the spaghetti code in order to make it really work by only using the `ConfigService`

File: packages/relay/src/lib/services/hbarLimitService/index.ts [hunk 9]
```diff
@@ -74,30 +77,37 @@ export class HbarLimitService implements IHbarLimitService {
...
-    this.remainingBudget = this.totalBudget;
 
-    if (this.totalBudget.toTinybars().lte(0)) {
+    const operatorId = ConfigService.get('OPERATOR_ID_MAIN');
+    const operatorKey = ConfigService.get('OPERATOR_KEY_MAIN');
+    if (operatorId) {
+      this.operatorAddress = AccountId.fromString(operatorId as string).toSolidityAddress();
+    } else if (operatorKey) {
+      this.operatorAddress = Utils.createPrivateKeyBasedOnFormat(operatorKey as string).publicKey.toEvmAddress();
+    }
```
[the remark this answers is not in the document]
Comment (victor-yanev) [reply packages/relay/src/lib/services/hbarLimitService/index.ts:107]: Yeah, that's true but in some cases the `OPERATOR_ID` is set to an empty string (e.g., in the `charts:install` workflow)

File: packages/relay/src/lib/db/types/hbarLimiter/subscriptionTier.ts (hunk 4 above)
Comment (victor-yanev) [reply packages/relay/src/lib/db/types/hbarLimiter/subscriptionTier.ts:40]: I suggest that we do it in a separate PR to not clutter this one even more with changes

File: packages/relay/src/lib/services/hbarLimitService/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (quiet-node) [reply packages/relay/src/lib/services/hbarLimitService/index.ts:563]: Nice thanks for the explanation! It makes sense!

File: packages/relay/src/lib/services/hbarLimitService/index.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (quiet-node) [reply packages/relay/src/lib/services/hbarLimitService/index.ts:107]: Hmm I thought the ConfigService would reject if OPERATOR_ID_MAIN is not provided? I was looking in the chars:install workflow and couldn't see where it set to empty. Might have missed it or might not be easy to see. Could you kindly share a link to where it set to empty string just wanted to see how it works. Thanks!

File: packages/server/tests/acceptance/hbarLimiter.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (quiet-node) [reply packages/server/tests/acceptance/hbarLimiter.spec.ts:269]: seems like a duplicated comment. Thanks for pointing it out tho!

File: docs/configuration.md (hunk 3 above)
Comment (Nana-EC) [reply docs/configuration.md:114]: Hmmm, we can't remove configs easily cause we may have operators who've adopted it.
Unless it's removal makes no difference to an operator we should leave it in an have a deprecation plan

File: packages/relay/src/lib/db/types/hbarLimiter/subscriptionTier.ts (hunk 4 above)
Comment (Nana-EC) [reply packages/relay/src/lib/db/types/hbarLimiter/subscriptionTier.ts:40]: Agreed

File: docs/configuration.md (hunk 3 above)
Comment (victor-yanev) [reply docs/configuration.md:114]: Well, in this case it makes no difference because this configuration is no longer used. 

I guess we should note in the release notes it was deprecated (it was used in the old hbar limiter, but we didn't include support for it in the redesigned version)

File: packages/server/tests/acceptance/hbarLimiter.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (victor-yanev) [reply packages/server/tests/acceptance/hbarLimiter.spec.ts:269]: I added assertions in the tests for `addExpense` in `hbarLimitService.spec.ts`:
```typescript
      const addExpensePromise = hbarLimitService.addExpense(
        expense,
        evmAddress,
        new RequestDetails({
          ...requestDetails,
          ipAddress,
        }),
      );

      let operatorPlan: IDetailedHbarSpendingPlan;
      if (evmAddress) {
        await expect(addExpensePromise).to.eventually.be.fulfilled;
        sinon.assert.calledTwice(hbarSpendingPlanRepositorySpy.addToAmountSpent); // once for operator and once for evm address
        operatorPlan = await hbarLimitService['getOperatorSpendingPlan'](requestDetails);
        sinon.assert.calledWith(hbarSpendingPlanRepositorySpy.addToAmountSpent.firstCall, operatorPlan.id, expense);
        sinon.assert.calledWith(hbarSpendingPlanRepositorySpy.addToAmountSpent.secondCall, sinon.match.string, expense);
        await Promise.all(updateAverageAmountSpentPerSubscriptionTierSpy.returnValues);
        const expectedAverageUsage = Math.round((otherPlanOfTheSameTier.amountSpent + expense) / 2);
        sinon.assert.calledOnceWithExactly(setAverageSpendingPlanAmountSpentGaugeSpy, expectedAverageUsage);
        sinon.assert.calledOnceWithExactly(incUniqueSpendingPlansCounterSpy, 1);
      } else {
        await expect(addExpensePromise).to.eventually.be.fulfilled;
        sinon.assert.calledWith(
          loggerSpy.warn,
          `${requestDetails.formattedRequestId} Cannot add expense to a spending plan without an evm address`,
        );
        sinon.assert.calledOnce(hbarSpendingPlanRepositorySpy.addToAmountSpent); // only once for the operator
        operatorPlan = await hbarLimitService['getOperatorSpendingPlan'](requestDetails);
        sinon.assert.calledWith(hbarSpendingPlanRepositorySpy.addToAmountSpent.firstCall, operatorPlan.id, expense);
      }
      sinon.assert.calledWith(hbarSpendingPlanRepositorySpy.addToAmountSpent, sinon.match.string, expense);
      expect((await hbarLimitService['getRemainingBudget'](requestDetails)).toTinybars().toNumber()).to.eq(
        totalBudgetInTinybars - expense,
      );
      expect((await hbarLimitService['hbarLimitRemainingGauge'].get()).values[0].value).to.equal(
        totalBudgetInTinybars - expense,
      );
      expect(operatorPlan.amountSpent).to.eq(expense);
```

[the remark this answers is not in the document]
Comment (quiet-node) [reply packages/server/tests/acceptance/hbarLimiter.spec.ts:269]: Nice! That's all I needed to make sure! Thanks for confirming!

File: packages/relay/src/lib/relay.ts (hunk 1 above)
Comment (quiet-node) [reply packages/relay/src/lib/relay.ts:170]: Sorry gotta unresolve this convo for easier tracking. Would this be considered technical debt that will be removed (if the logic is done in the service class) later when the draft PR is ready?

File: packages/relay/src/lib/services/hapiService/hapiService.ts [hunk 10]
```diff
@@ -289,6 +290,9 @@ export default class HAPIService {
     this.clientMain = this.initClient(this.logger, this.hederaNetwork);
     this.client = this.initSDKClient(this.logger);
     this.resetCounters();
+    if (this.clientMain.operatorAccountId) {
+      this.hbarLimitService.setOperatorAddress(this.clientMain.operatorAccountId.toSolidityAddress());
+    }
```
Comment (quiet-node) [inline packages/relay/src/lib/services/hapiService/hapiService.ts:295]: same question as above, would this be considered technical debt that will be removed (if the logic is done in the service class) later when the draft PR is ready?

**Review by quiet-node — APPROVED** [review]
Just couple of Qs otherwise everything LGTM! Great work! 🚀

File: packages/relay/src/lib/relay.ts (hunk 1 above)
Comment (victor-yanev) [reply packages/relay/src/lib/relay.ts:170]: It's more like a workaround due to a technical debt coming from [this PR](https://github.com/hashgraph/hedera-json-rpc-relay/pull/3021/)

File: packages/relay/src/lib/services/hbarLimitService/index.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (victor-yanev) [reply packages/relay/src/lib/services/hbarLimitService/index.ts]: `required` is set to `true` in the `ConfigService`, but some workflows just pass an empty string to bypass this required constraint, same for the acceptance tests - empty string is passed from `{{ input.operator_id }}` and then later on the actual value from the parsed `.env` file is being used (but it's actually not set as an environment variable - see https://github.com/hashgraph/hedera-json-rpc-relay/pull/3021/)

**Review by konstantinabl — APPROVED** [review]
